internal/cli/resolve.go

```diff
@@ -0,0 +1,251 @@
+			}
+
+			if len(logPolicy) == 0 {
+				if quietEnabled {
+					logPolicy = []string{"builtin:discard"}
+				} else {
+					logPolicy = []string{"builtin:stderr"}
+				}
+			}
+
+			logWriter, err := iocomb.Combine(logPolicy)
+			if err != nil {
+				return fmt.Errorf("invalid logging policy: %w", err)
+			}
+			logger := log.NewLogger(logWriter)
+
+			archs := types.ParseArchitectures(archstrs)
+
+			return ResolveCmd(
+				cmd.Context(),
+				output,
+				archs,
+				[]build.Option{
+					build.WithLogger(logger),
+					build.WithConfig(args[0]),
+					build.WithExtraKeys(extraKeys),
+					build.WithExtraRepos(extraRepos),
+					build.WithDebugLogging(debugEnabled),
+				},
+			)
+		},
+	}
+
+	cmd.Flags().StringSliceVarP(&extraKeys, "keyring-append", "k", []string{}, "path to extra keys to include in the keyring")
+	cmd.Flags().StringSliceVarP(&extraRepos, "repository-append", "r", []string{}, "path to extra repositories to include")
+	cmd.Flags().StringSliceVar(&archstrs, "arch", nil, "architectures to build for (e.g., x86_64,ppc64le,arm64) -- default is all, unless specified in config. Can also use 'host' to indicate arch of host this is running on")
+	cmd.Flags().StringVar(&output, "output", "", "path to file where lock file will be written")
+	cmd.Flags().StringSliceVar(&logPolicy, "log-policy", []string{}, "logging policy to use")
+	cmd.Flags().BoolVar(&debugEnabled, "debug", false, "enable debug logging")
+	cmd.Flags().BoolVar(&quietEnabled, "quiet", false, "disable logging")
+
+	return cmd
+}
+
+func ResolveCmd(ctx context.Context, output string, archs []types.Architecture, opts []build.Option) error {
+	wd, err := os.MkdirTemp("", "apko-*")
+	if err != nil {
+		return fmt.Errorf("failed to create working directory: %w", err)
+	}
+	defer os.RemoveAll(wd)
+
+	o, ic, err := build.NewOptions(opts...)
+	if err != nil {
+		return err
+	}
+
+	// cases:
+	// - archs set: use those archs
+	// - archs not set, bc.ImageConfiguration.Archs set: use Config archs
+	// - archs not set, bc.ImageConfiguration.Archs not set: use all archs
+	switch {
+	case len(archs) != 0:
+		ic.Archs = archs
+	case len(ic.Archs) != 0:
+		// do nothing
+	default:
+		ic.Archs = types.AllArchs
+	}
+	// save the final set we will build
+	archs = ic.Archs
+	o.Logger().Infof("Determining packages for %d architectures: %+v", len(ic.Archs), ic.Archs)
+
+	// The build context options is sometimes copied in the next functions. Ensure
+	// we have the directory defined and created by invoking the function early.
+	defer os.RemoveAll(o.TempDir())
+
+	lock := lock{
+		Version: "v1",
+		Contents: lockContents{
+			Packages:     []lockPkg{},
+			Repositories: []lockRepo{},
+			Keyrings:     []lockKeyring{},
+		},
+	}
+
+	repositories := map[string]bool{}
+
+	for _, keyring := range ic.Contents.Keyring {
+		lock.Contents.Keyrings = append(lock.Contents.Keyrings, lockKeyring{
+			Name: stripURLScheme(keyring),
+			URL:  keyring,
+		})
+	}
+
+	for _, arch := range archs {
```

Comment:
```suggestion
	// TODO: If the archs can't agree on package versions (e.g., arm builds are ahead of x86) then we should fail instead of producing inconsistent locks.
	for _, arch := range archs {
```